Engine Disposal

gianchub opened the pull request:
@mmcardle 

This is a fix in response to https://github.com/gianchub/sqlalchemy-diff/issues/32.

It:
- Adds the ability to dispose an engine, which closes the connection pool currently used and creates a new one immediately. (See [here](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Engine.dispose))
- Adds tests for new functionality.
- Updates the README
- Updates scaffolding, versions, requirements, etc.

gianchub committed:
Enhance Comparer class with engine disposal feature

- Updated the Comparer class to include a `dispose_engines` flag for automatic disposal of database engines after comparison.
- Modified the `from_params` classmethod to set this flag to true by default.
- Added a `dispose` method to close pooled connections.
- Updated README.md to reflect these changes and provide usage notes.
- Introduced tests to verify engine disposal behavior.

gianchub committed:
Update requirements, version and changelog.

gianchub committed:
Better README.

mmcardle commented on src/sqlalchemydiff/comparer.py:
```diff
@@ -157,3 +169,8 @@ def _get_db_info(
+
+    def dispose(self) -> None:
+        """Dispose engines to close any pooled connections."""
+        self.db_one_engine.dispose()
```
Can dispose fail on `db_one_engine` and leave `db_two_engine` undisposed?
Would it be useful to have both in a try...catch

gianchub replied:
Reading the source code for it, it ends up calling `pool.dispose()` which should safely dispose of the object. I think the likelihood this blows up is very remote. But even if it blows up, then disposing at least the other one won't still solve the issue. My personal feeling is that it should be good enough, but happy to accept suggestions :)

vkruglik-aka replied:
Looks good to me - thank you!

mmcardle commented on tests/test_comparer.py:
```diff
@@ -158,6 +159,39 @@ def test_dump_result(
+        comparer.compare()
+
+        engine_one.dispose.assert_called_once_with()
+        engine_two.dispose.assert_called_once_with()
```
This could both be assert_called_once()
https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once

gianchub replied:
both fixed (there was another pair in another test), thank you for spotting it

gianchub committed:
Refactor test assertions for engine calls

- Updated assertions in `test_comparer.py` to use `assert_called_once()` instead of `assert_called_once_with()` for better clarity and consistency in testing engine calls.

mmcardle reviewed (dismissed):
LGTM 👍

gianchub committed:
Refactor Comparer class engine disposal methods

- Changed the private attribute for engine disposal from `_dispose_engines` to `__dispose_engines` for better encapsulation.
- Updated the disposal logic in the `finally` block to call the renamed `_dispose_engines` method, ensuring proper closure of pooled connections.

gianchub committed:
Refactor database engine fixtures to ensure proper disposal

- Updated the `db_engine_one` and `db_engine_two` fixtures in `base.py` to use a context manager for engine disposal, ensuring that resources are properly released after use.
- Modified the `sqlite_db_engine_one` and `sqlite_db_engine_two` fixtures in `test_comparer.py` similarly to enhance resource management during tests.

mmcardle reviewed (approved):
LGTM still👍
